src/core/ucc_coll.c

```diff
@@ -223,23 +224,167 @@ ucc_status_t ucc_collective_post(ucc_coll_req_h request)
-    ucc_coll_task_t *task = ucc_derived_of(ev->req, ucc_coll_task_t);
+    ucc_coll_task_t *task = ucc_derived_of(request, ucc_coll_task_t);
+
+    ucc_debug("coll_finalize: req %p, seq_num %u", task, task->seq_num);
+    return task->finalize(task);
+}
+
+static ucc_status_t ucc_triggered_task_finalize(ucc_coll_task_t *task)
+{
+    ucc_trace("finalizing triggered ev task %p", task);
+    ucc_free(task);
+    return UCC_OK;
+}
+
+static ucc_status_t ucc_triggered_coll_complete(ucc_coll_task_t *parent_task, //NOLINT
+                                                ucc_coll_task_t *task)
+{
+    ucc_trace("triggered collective complete, task %p, seq_num %u",
+              task, task->seq_num);
+    return ucc_mc_ee_task_end(task->ee_task, task->ee->ee_type);
+}
+
+static ucc_status_t ucc_trigger_complete(ucc_coll_task_t *parent_task,
+                                         ucc_coll_task_t *task)
+{
+    ucc_status_t status;
+
+    ucc_trace("event triggered, ev_task %p, coll_task %p, seq_num %u",
+              parent_task, task, task->seq_num);
+
+    task->ee_task = parent_task->ee_task;
+    status        = task->post(task);
+    if (ucc_unlikely(status != UCC_OK)) {
+        ucc_error("failed to post triggered coll, task %p, seq_num %u, %s",
+                  task, task->seq_num, ucc_status_string(status));
+        return status;
+    }
+
+    if (task->super.status == UCC_OK) {
+        return ucc_triggered_coll_complete(task, task);
+    } else {
+        ucc_assert(task->super.status == UCC_INPROGRESS);
+        if (task->ee_task) {
+            // TODO use CB instead of EM
+            ucc_event_manager_subscribe(&task->em, UCC_EVENT_COMPLETED, task,
+                                        ucc_triggered_coll_complete);
+        }
+    }
+    return UCC_OK;
+}
+
+static ucc_status_t ucc_trigger_test(ucc_coll_task_t *task)
+{
+    ucc_status_t status;
+    ucc_ev_t    *post_event;
+    ucc_ev_t    *ev;
+
+
+    if (task->ev == NULL) {
+        if (task->ee->ee_type == UCC_EE_CUDA_STREAM) {
+            /* implicit event triggered */
+            task->ev = (ucc_ev_t *) 0xFFFF; /* dummy event */
+            task->ee_task = NULL;
+        } else if (UCC_OK == ucc_ee_get_event_internal(task->ee, &ev,
+                                                 &task->ee->event_in_queue)) {
+            ucc_trace("triggered event arrived, ev_task %p", task);
+            task->ev      = ev;
+            task->ee_task = NULL;
+        } else {
+            return UCC_OK;
+        }
+    }
+
+    if (task->ee_task == NULL) {
+        status = ucc_mc_ee_task_post(task->ee->ee_context,
+                                     task->ee->ee_type, &task->ee_task);
+        if (ucc_unlikely(status != UCC_OK)) {
+            ucc_error("error in ee task post, %s", ucc_status_string(status));
+            task->super.status = status;
+            return status;
+        }
+
+        /* TODO: mpool */
+        post_event = ucc_malloc(sizeof(ucc_ev_t), "event");
```

Comment:
when post_event will be freed?

Reply:
the consumer (ucc_ee_get_event) should call ucc_ee_ack_event

Reply:
yes, i did that